File: skills/saki/scripts/context_scanner.js

/**
 * Saki Context Scanner (Node.js Fallback)
 * Zero-dependency script to analyze project context.
 */
const fs = require('fs');
const path = require('path');

function scanContext(rootDir = process.cwd()) {
    const context = {
        files: [],
        tech_stack: [],
        frameworks: [],
        description: "",
        package_manager: "unknown",
        is_git_repo: fs.existsSync(path.join(rootDir, ".git"))
    };

    try {
        const files = fs.readdirSync(rootDir);
        const ignored = [".git", "node_modules", ".venv", "__pycache__", "target", "dist", "build"];
        context.files = files.filter(f => !ignored.includes(f));

        // 1. Node.js
        if (files.includes("package.json")) {
            context.tech_stack.push("node");
            try {
                const pkgPath = path.join(rootDir, "package.json");
                const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
                const deps = { ...pkg.dependencies, ...pkg.devDependencies };

                if (deps["react"]) context.frameworks.push("react");
                if (deps["next"]) context.frameworks.push("nextjs");
                if (deps["vue"]) context.frameworks.push("vue");
                if (deps["typescript"]) context.tech_stack.push("typescript");
                if (deps["tailwindcss"]) context.frameworks.push("tailwindcss");
            } catch (e) {}

            if (files.includes("yarn.lock")) context.package_manager = "yarn";
            else if (files.includes("pnpm-lock.yaml")) context.package_manager = "pnpm";
            else if (files.includes("package-lock.json")) context.package_manager = "npm";
        }

        // 2. Python
        const pyFiles = ["requirements.txt", "pyproject.toml", "setup.py", "Pipfile"];
        if (files.some(f => pyFiles.includes(f))) {
            context.tech_stack.push("python");
            if (files.includes("manage.py")) context.frameworks.push("django");
            if (files.includes("app.py")) context.frameworks.push("flask"); // Simple heuristic
            if (files.includes("main.py")) context.frameworks.push("fastapi"); // Simple heuristic
        }

        // 3. Rust
        if (files.includes("Cargo.toml")) {
            context.tech_stack.push("rust");
            try {
                const cargo = fs.readFileSync(path.join(rootDir, "Cargo.toml"), 'utf8');
                if (cargo.includes("tokio")) context.frameworks.push("tokio");
                if (cargo.includes("actix")) context.frameworks.push("actix");
                if (cargo.includes("tauri")) context.frameworks.push("tauri");
            } catch (e) {}
        }
        
        // 4. README Analysis
        if (files.includes("README.md")) {
            try {
                const readme = fs.readFileSync(path.join(rootDir, "README.md"), 'utf8');
                 context.description = readme.slice(0, 200).replace(/\n/g, " ").trim();
            } catch (e) {}
        }

    } catch (e) {
        context.error = e.message;
    }

    // Deduplicate
    context.tech_stack = [...new Set(context.tech_stack)];
    context.frameworks = [...new Set(context.frameworks)];

    console.log(JSON.stringify(context, null, 2));
}

// Run
scanContext();

File: skills/saki/scripts/context_scanner.py

import os
import json
import sys
import re

def analyze_readme(root_dir):
    """
    Scans README.md for keywords to mitigate missing package files.
    """
    readme_path = os.path.join(root_dir, "README.md")
    keywords = []
    description = ""
    
    if os.path.exists(readme_path):
        try:
            with open(readme_path, "r", encoding="utf-8") as f:
                content = f.read()
                description = content[:200].replace("\n", " ").strip()
                
                # Heuristic keyword extraction
                triggers = {
                    "django": "django", "flask": "flask", "fastapi": "fastapi",
                    "react": "react", "vue": "vue", "next.js": "nextjs", "nextjs": "nextjs",
                    "pytorch": "pytorch", "tensorflow": "tensorflow",
                    "docker": "docker", "kubernetes": "kubernetes",
                    "rust": "rust", "tokio": "tokio", "actix": "actix",
                    "go": "go", "gin": "gin", "chi": "chi"
                }
                
                normalized_content = content.lower()
                for key, val in triggers.items():
                    if key in normalized_content:
                        keywords.append(val)
        except Exception:
            pass
            
    return list(set(keywords)), description

def scan_context(root_dir="."):
    context = {
        "files": [],
        "tech_stack": [],
        "frameworks": [],
        "description": "",
        "package_manager": "unknown",
        "is_git_repo": os.path.isdir(os.path.join(root_dir, ".git"))
    }
    
    try:
        files = os.listdir(root_dir)
        context["files"] = [f for f in files if f not in [".git", "node_modules", ".venv", "__pycache__", "target", "dist", "build"]]
        
        # 1. Node.js
        if "package.json" in files:
            context["tech_stack"].append("node")
            try:
                with open(os.path.join(root_dir, "package.json"), "r") as f:
                    pkg = json.load(f)
                    deps = {**pkg.get("dependencies", {}), **pkg.get("devDependencies", {})}
                    
                    if "react" in deps: context["frameworks"].append("react")
                    if "next" in deps: context["frameworks"].append("nextjs")
                    if "vue" in deps: context["frameworks"].append("vue")
                    if "typescript" in deps: context["tech_stack"].append("typescript")
                    if "tailwindcss" in deps: context["frameworks"].append("tailwindcss")
                    
            except:
                pass
            
            if "yarn.lock" in files: context["package_manager"] = "yarn"
            elif "pnpm-lock.yaml" in files: context["package_manager"] = "pnpm"
            elif "package-lock.json" in files: context["package_manager"] = "npm"

        # 2. Python
        if any(f in files for f in ["requirements.txt", "pyproject.toml", "setup.py", "Pipfile"]):
            context["tech_stack"].append("python")
            
            # Simple content check for common frameworks
            py_indicators = {
                "django": ["manage.py", "wsgi.py"],
                "flask": ["app.py", "wsgi.py"],
                "fastapi": ["main.py"]
            }
            
            for fw, indicators in py_indicators.items():
                if any(ind in files for ind in indicators):
                    context["frameworks"].append(fw)

        # 3. Rust
        if "Cargo.toml" in files:
            context["tech_stack"].append("rust")
            try:
                with open(os.path.join(root_dir, "Cargo.toml"), "r", encoding="utf-8") as f:
                    content = f.read()
                    if "tokio" in content: context["frameworks"].append("tokio")
                    if "actix" in content: context["frameworks"].append("actix")
                    if "axum" in content: context["frameworks"].append("axum")
                    if "tauri" in content: context["frameworks"].append("tauri")
            except:
                pass

        # 4. Go
        if "go.mod" in files:
            context["tech_stack"].append("go")
            try:
                with open(os.path.join(root_dir, "go.mod"), "r", encoding="utf-8") as f:
                    content = f.read()
                    if "gin-gonic" in content: context["frameworks"].append("gin")
                    if "gofiber" in content: context["frameworks"].append("fiber")
            except:
                pass
        
        # 5. README Analysis (Augmentation)
        readme_keywords, readme_desc = analyze_readme(root_dir)
        context["description"] = readme_desc
        
        # Merge README keywords into frameworks if not already present
        for kw in readme_keywords:
            if kw not in context["frameworks"] and kw not in context["tech_stack"]:
                context["frameworks"].append(kw)

    except Exception as e:
        context["error"] = str(e)

    # De-duplicate
    context["tech_stack"] = list(set(context["tech_stack"]))
    context["frameworks"] = list(set(context["frameworks"]))
    
    return context

if __name__ == "__main__":
    result = scan_context(os.getcwd())
    print(json.dumps(result, indent=2))
